pkg/apis/batch/validation/validation.go

```diff
@@ -389,6 +399,92 @@ func validatePodFailurePolicyRuleOnExitCodes(onExitCode *batch.PodFailurePolicyO
 	return allErrs
 }
 
+func validateSuccessPolicy(spec *batch.JobSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	criteriaPath := fldPath.Child("criteria")
+	if len(spec.SuccessPolicy.Criteria) == 0 {
+		allErrs = append(allErrs, field.Required(criteriaPath, "at least one criteria must be specified when the successPolicy is specified"))
+	}
+	for i, criterion := range spec.SuccessPolicy.Criteria {
+		allErrs = append(allErrs, validateSuccessPolicyCriteria(spec, &criterion, criteriaPath.Index(i))...)
+	}
+	return allErrs
+}
+
+func validateSuccessPolicyCriteria(spec *batch.JobSpec, criteria *batch.SuccessPolicyCriteria, criteriaPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if criteria.SucceededCount == nil && criteria.SucceededIndexes == nil {
+		allErrs = append(allErrs, field.Required(criteriaPath, "must be specified succeededCount or succeededIndexes"))
+	}
+	var totalIndexes *int
+	if criteria.SucceededIndexes != nil {
+		succeededIndexes := criteriaPath.Child("succeededIndexes")
+		if len(*criteria.SucceededIndexes) > maxJobSuccessPolicySucceededIndexesLimit {
+			allErrs = append(allErrs, field.TooLong(succeededIndexes, *criteria.SucceededIndexes, maxJobSuccessPolicySucceededIndexesLimit))
+		}
+		var err error
+		if totalIndexes, err = validateIndexesFormat(*criteria.SucceededIndexes, int(*spec.Completions)); err != nil {
+			allErrs = append(allErrs, field.Invalid(succeededIndexes, *criteria.SucceededIndexes, fmt.Sprintf("error parsing succeededIndexes: %s", err.Error())))
+		}
+	}
+	if criteria.SucceededCount != nil {
+		succeededCountPath := criteriaPath.Child("succeededCount")
+		if *criteria.SucceededCount < 0 {
+			allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(*criteria.SucceededCount), succeededCountPath)...)
+		}
+		if *criteria.SucceededCount > *spec.Completions {
+			allErrs = append(allErrs, field.Invalid(succeededCountPath, *criteria.SucceededCount, fmt.Sprintf("must be less than or equal to %d (the number of specified completions)", *spec.Completions)))
+		}
+		if totalIndexes != nil && int(*criteria.SucceededCount) > *totalIndexes {
+			allErrs = append(allErrs, field.Invalid(succeededCountPath, *criteria.SucceededCount, fmt.Sprintf("must be less than or equal to %d (the number of indexes in the specified succeededIndexes field)", *totalIndexes)))
+		}
+	}
+	return allErrs
+}
+
+// Extended the [validateIndexesFormat](https://github.com/kubernetes/kubernetes/pull/123273)
+// TODO: Once #123273 is merged into the master, we should rebase these commits.
+func validateIndexesFormat(indexesStr string, completions int) (*int, error) {
```

Comment:
After #123273 is merged, I will rebase this PR and re-organize this function a little bit.

Reply:
Reminding myself that this PR is contingent on https://github.com/kubernetes/kubernetes/pull/123273